llvm/test/Instrumentation/AddressSanitizer/remove-memory-effects.ll

```diff
@@ -0,0 +1,23 @@
+; Remove possible memory effects from functions that are invalidated by
+; AddressSanitizer instrumentation.
+
+; RUN: opt < %s -passes=asan -asan-use-after-scope -S | FileCheck %s
+
+target datalayout = "e-i64:64-f80:128-s:64-n8:16:32:64-S128"
+target triple = "x86_64-unknown-linux-gnu"
+
+; CHECK: @foo(ptr) #[[ATTRS_FOO:[0-9]+]]
+declare void @foo(ptr) memory(argmem: write)
+
+; CHECK: @bar() #[[ATTRS_BAR:[0-9]+]]
+define void @bar() sanitize_address {
+entry:
+  %x = alloca i32, align 4
+  call void @llvm.lifetime.start.p0(i64 4, ptr %x)
+  call void @foo(ptr %x)
+  call void @llvm.lifetime.end.p0(i64 4, ptr %x)
+  ret void
+}
+
```

Comment:
test the individual pointer argument removing writeonly case?

Reply:
WDYM?

Reply:
Arguments could have `writeonly` attributes, which `removeFnAttributes` drops. This is not covered by the test and would be good to add coverage, possibly in a different function .

Reply:
AFAIK, ASan does not read/write into the variables themselves and just stores metadata. Changed so only HWAddressSanitizer removes `writeonly` from specific arguments.

Reply:
Still trying to wrap my head around this, I'm fresh to controlling optimizations by attributes

> so only HWAddressSanitizer removes writeonly

I feel uneasy with this - so is this change addressing the problem w. HWASAN , but not with ASAN, when the write only attribute is present? Or the opposite (the number of negations in the removal condition overwhelms me a bit, as the original `The API is `weird` comment suggests)?

In other words - if there's a function compiled with  (not HW) ASAN instrumentation, decorated with writeonly attribute, will it still be dropped?

Reply:
It's a portion from common code which ASan does not require. HWASan, however, when short granules are used, might read from the memory pointed by via pointer arguments.
To summarize, ASan will not drop `writeonly`.

Reply:
I've also made it so HWASan drops `writeonly` only when `UseShortGranules` is true.